test/e2e/util.go

```diff
@@ -154,51 +160,65 @@ func conditionsMatchExpected(expected, actual map[string]string) bool {
 	return reflect.DeepEqual(expected, filtered)
 }
 
-func defaultExternalDNS(name, zoneID, zoneDomain string) operatorv1alpha1.ExternalDNS {
-	return operatorv1alpha1.ExternalDNS{
+func defaultExternalDNS(name, zoneID, zoneDomain, sourceType, routerName string) operatorv1alpha1.ExternalDNS {
+	resource := operatorv1alpha1.ExternalDNS{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: name,
 		},
 		Spec: operatorv1alpha1.ExternalDNSSpec{
 			Zones: []string{zoneID},
-			Source: operatorv1alpha1.ExternalDNSSource{
-				ExternalDNSSourceUnion: operatorv1alpha1.ExternalDNSSourceUnion{
-					Type: operatorv1alpha1.SourceTypeService,
-					Service: &operatorv1alpha1.ExternalDNSServiceSourceOptions{
-						ServiceType: []corev1.ServiceType{
-							corev1.ServiceTypeLoadBalancer,
-							corev1.ServiceTypeClusterIP,
-						},
-					},
-					AnnotationFilter: map[string]string{
-						"external-dns.mydomain.org/publish": "yes",
-					},
+		},
+	}
+
+	switch sourceType {
+	case "Service":
+		resource.Spec.Source = getServiceTypeSource(zoneDomain)
+	case "OpenShiftRoute":
+		resource.Spec.Source = getOpenshiftRouteTypeSource(routerName, zoneDomain)
+		// TODO : test with default as routername for default router
+		if routerName == "" || routerName == "default" {
+			resource.Spec.Source.OpenShiftRoute = nil
+		}
+	default:
+		//TODO : we should return error and fail the test on Source type
+		fmt.Errorf("source type not support : %s", sourceType)
```

Comment:
Why not making 2 functions: `testExternalDNSServiceSource` and `testExternalDNSRouteSource` and not bother about passing the right source type (in the right case too)?
This would make the function call lighter too: 1 parameter fewer.

Reply:
there is no issue, we are write two functions,
but the we loose the code re-usability.

Reply:
these function call will be replaced by inline. actual calling will be optimized by go itself.

Reply:
The usability will not be harmed, you will still use the same common base and will just add the source `Service` for `testExternalDNSServiceSource` and the source `Route` for `testExternalDNSRouteSource`. Similar to how we build `ExternalDNS` instance in the helpers.
This will avoid the delema about  if the source given is not `Source` or `OpenShiftRoute`.

Reply:
sure, done.